### ISSUE TITLE
Bug in closest.js?

### ISSUE THREAD

**Issue by riosn-n** [msn0/mdn-polyfills#49]
In line 3, document.documentElement.contains only succeeds if the element is present in window.document. For a document instance parsed from a string using DOMParser.parseFromString it won't work. Is there a reason to have this check?

**Comment by msn0** [comment]
Could you provide some more details, e.g. a test code that fails?

**Comment by idmadj** [comment]
Line 3 seems to be there to emulate [Edge's behavior of returning `null` if the element isn't added to the DOM](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest#Compatibility_notes).

**Comment by riosn-n** [comment]
Sorry for not providing more details initially. If you call [DOMParser.parseFromString](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) it returns a new instance of Document that is different from window.document. 

If I now get an element from that new document and call .closest() on it, when the polyfill executes line 3, it will return null:

    if (!document.documentElement.contains(el)) return null;

because it is checking if my element is contained in window.document, which is not true. My element is an element of the other document created by the DOMParser. So line 3 should probably try to achieve that effect by working with el.document and not window.document.

Does that help?

**Comment by idmadj** [comment]
Ultimately this about choosing whether the polyfill should behave like Edge's implementation, or simply stick to the spec which makes no mention of requiring the element to be attached to the DOM.

Your use case seems legitimate and I'm leaning towards following the spec, but removing the check would be a breaking change.

Thoughts @msn0 ?

**Comment by riosn-n** [comment]
@idmadj, the polyfill's implementation is not correct.

If you run this in Edge, it will work:

var doc = (new DOMParser()).parseFromString('<parent><child id="child"></child></parent>', 'text/xml');
var childElement = doc.getElementById("child");
console.log(childElement.closest('parent')); // returns parent element

However, the polyfill would incorrectly fail in this case (in line 3) because "doc" is not "window.document".

**Comment by idmadj** [comment]
Seems like Edge has very inconsistent behavior @riosn-n. 

Funnily enough, running this in Edge:
```
document.createElement("div").closest("div");
```
returns `null`.

In other words, `closest` works with an unattached element generated by `DOMParser`, but fails with `createElement`.

Not having to deal with these quirks seems like another good reason to stick to spec IMHO.

**Comment by riosn-n** [comment]
Setting Edge aside for a moment, the polyfill fails with the code I provided in all other browsers. I believe we can all agree that's a bug.

In the case of Edge, the polyfill also fails to do what the browser does. The whole point of line 3 was to imitate Edge, but it is only imitating it partially. The question is: can we make a small change in the polyfill so that 1) the bug is no longer present in all other browsers, and 2) in the case of Edge it maintains the special behavior while not failing in a document generated using DOMParser? I believe it is possible... and by looking [here](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest#Polyfill) it seems that maybe the code in this library no longer reflects the latest version of their snippet? I don't see the line that creates the problem there anymore.

**Comment by alexkrolick** [comment]
It looks like IE11 only supports the `ms` prefixed version of matches, so the polyfill doesn't work there. I think it's also the only common browser that doesn't support `.closest` natively 😞 .

Latest MDN polyfill: https://developer.mozilla.org/en-US/docs/Web/API/Element/closest

<details>

```js
if (!Element.prototype.matches) {
  Element.prototype.matches = Element.prototype.msMatchesSelector || 
                              Element.prototype.webkitMatchesSelector;
}

if (!Element.prototype.closest) {
  Element.prototype.closest = function(s) {
    var el = this;

    do {
      if (el.matches(s)) return el;
      el = el.parentElement || el.parentNode;
    } while (el !== null && el.nodeType === 1);
    return null;
  };
}
```

</details>

**Comment by msn0** [comment]
I'll update the code to the newest MDN revision.

**Comment by msn0** [comment]
I've published  📦 mdn-polyfills@5.16.3 which contains the newest MDN version of Element.prototype.closest. Please check whether it works for you @riosn-n @idmadj @alexkrolick
